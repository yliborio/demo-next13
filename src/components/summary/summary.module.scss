@import "./../../styles/variables.scss";

.container {
  display: flex;
  flex-direction: column;
  background-color: $bg-secondary;
  border-radius: 8px;
  font-size: $font-size-2;
  font-weight: 600;
  padding: 2rem;

  justify-content: center;
  align-items: center;
  .button {
    font-size: $font-size-2;
    padding: 1rem 2rem;
    text-transform: uppercase;
    color: $bg-secondary;
    background-color: black;
    border: none;
  }
}
