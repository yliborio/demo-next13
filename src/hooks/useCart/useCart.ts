"use client";

import { FakeAPIProduct } from "core/types/product";
import { CartItems } from "core/types/cartItem";
import { useContext } from "react";
import { LocalContext } from "core/contexts/local-context";
import currency from "currency.js";

export const useCart = () => {    

    const {cart, updateCart} = useContext(LocalContext);    
    const {items, total} = cart;


    const getCartProducts = async () => {
        const data = await fetch(`https://fakestoreapi.com/products`);
        
        const idToQtdMap: Record<number, number> = {};
        for (const item of items) {
            idToQtdMap[item.id] = item.quantity;
        }
        const productData : FakeAPIProduct[]= await data.json();
        return  productData
            .filter((p : FakeAPIProduct) => idToQtdMap.hasOwnProperty(p.id))
            .map((product) => ({ product, quantity: idToQtdMap[product.id]}))
            .sort((a,b) => a.product?.id - b?.product?.id);
    }

    const addProduct = (product : FakeAPIProduct) => {
        const storageProduct = items.find((item) => item.id === product.id );  
        if(storageProduct){            
            const newValue: CartItems =  { 
                items: [...items.filter((p) => p.id !== product.id ),
                    {id: product.id, quantity: storageProduct.quantity + 1}],                
                total: currency(total).add(product.price)
            }
            updateCart(newValue)            
        }
        else{
            const newValue: CartItems =  { 
                items: [...items, {id: product.id, quantity: 1}],                
                total: currency(total).add(product.price)
            }
            updateCart(newValue)
        }
    }

    const removeProduct = (product : FakeAPIProduct) => {
        const storageProduct = items.find((item) => item.id === product.id );  
        if(storageProduct){    
            const newValue: CartItems =  { 
                items: [...items.filter((p) => p.id !== product.id )],
                total: currency(total).subtract(product.price * storageProduct.quantity)
            }
            updateCart(newValue)            
        }       
    }

    const removeQuantity = (product : FakeAPIProduct) => {        
        const storageProduct = items.find((item) => item.id === product.id ); 
        if (storageProduct?.quantity === 1) return removeProduct(product);
        if(storageProduct){            
            const newValue: CartItems =  { 
                items: [...items.filter((p) => p.id !== product.id ), {id: product.id, quantity: storageProduct.quantity - 1}],                
                total: currency(total).subtract(product.price)
            }
            updateCart(newValue)                   
        }       
    }


    return {cart, addProduct, getCartProducts, removeProduct, removeQuantity}
}