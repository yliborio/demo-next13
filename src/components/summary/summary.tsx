"use client";
import { useCart } from "../../hooks/useCart/useCart";
import styles from "./summary.module.scss";

export const Summary = () => {
  const {
    cart: { total, items },
  } = useCart();

  const text = items?.length > 1 ? " products" : " product";
  return (
    <div className={styles["container"]} data-testid="summary">
      <span>{`Subtotal (${items.length} ${text}): `} </span>
      <span>{`$${total}`} </span>
    </div>
  );
};
