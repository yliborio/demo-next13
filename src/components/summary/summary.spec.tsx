import React from "react";
import { render, screen } from "@testing-library/react";
import { Summary } from "./summary";
import * as cartHook from "../../hooks/useCart/useCart";

jest.mock("../../hooks/useCart/useCart");
describe("Summary >1 items", () => {
  it("displays the correct subtotal", () => {
    jest.spyOn(cartHook, "useCart").mockReturnValue({
      cart: {
        total: 100,
        items: [{ id: 1 }, { id: 2 }],
      },
    } as any);

    const { asFragment } = render(<Summary />);
    expect(asFragment()).toMatchSnapshot();

    const subtotalElement = screen.getByText(/Subtotal/);
    expect(subtotalElement).toBeInTheDocument();

    const priceElement = screen.getByText("$100");
    expect(priceElement).toBeInTheDocument();
  });

  it("displays the product instead of products if items length == 1", () => {
    jest.spyOn(cartHook, "useCart").mockReturnValue({
      cart: {
        total: 100,
        items: [{ id: 1 }],
      },
    } as any);
    render(<Summary />);
    const productText = screen.getByText(/product/);
    expect(productText).toBeInTheDocument();
  });
});
