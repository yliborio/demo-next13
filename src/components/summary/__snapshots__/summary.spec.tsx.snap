// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary >1 items displays the correct subtotal 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="summary"
  >
    <span>
      Subtotal (2  products):  
    </span>
    <span>
      $100 
    </span>
  </div>
</DocumentFragment>
`;
